src/freenet/node/PeerLocation.java

```diff
@@ -107,4 +111,101 @@ synchronized double setLocation(double newLoc) {
 		return oldLoc;
 	}
 
+	/**
+	 * Finds the position of the first element in the sorted list greater than the given element,
+	 * or -1 if none.
+	 * This is a binary search of logarithmic complexity.
+	 */
+	private static int findFirstGreater(final double[] elems, final double x) {
+		int low = 0;
+		int high = elems.length;
+		int mid;
+		if (elems[high - 1] <= x) {
+			return -1;
+		}
+		while (low != high) {
+			mid = low + (high - low) / 2;
+			if (elems[mid] > x) {
+				high = mid;
+			} else {
+				low = mid + 1;
+			}
+		}
+		return low;
+	}
+
+	/**
+	 * Finds the position of the closest location in the sorted list of locations.
+	 * This is a binary search of logarithmic complexity.
+	 */
+	static int findClosestLocation(final double[] locs, final double l) {
+		assert(locs.length > 0);
+		if (locs.length == 1) {
+			return 0;
+		}
+		final int firstGreater = findFirstGreater(locs, l);
+		final int left;
+		final int right;
+		if (firstGreater == -1 || firstGreater == 0) {
+			// All locations are greater, or all locations are smaller.
+			// Closest location must be either smallest or greatest location.
+			left = locs.length - 1;
+			right = 0;
+		} else {
+			left = firstGreater - 1;
+			right = firstGreater;
+		}
+		if (Location.distance(l, locs[left]) <= Location.distance(l, locs[right])) {
+			return left;
+		}
+		return right;
+	}
+
+	/**
+	 * Finds the closest non-excluded peer in O(log n + m) time, where n is the number of peers and
+	 * m the number of exclusions.
+	 * @param exclude the set of locations to exclude, may be null
+	 * @return the closest non-excluded peer's location, or NaN if none is found
+	 */
+	public double getClosestPeerLocation(final double l, Set<Double> exclude) {
+		final double[] locs;
+		synchronized (this) {
+			locs = currentPeersLocation;
+		}
+		if (locs == null || locs.length == 0) {
+			return Double.NaN;
+		}
+		final int closest = findClosestLocation(locs, l);
+		if (exclude == null || !exclude.contains(locs[closest])) {
+			return locs[closest];
+		}
+
+		int left = (closest == 0) ? (locs.length - 1) : (closest - 1);
+		int right = (closest == locs.length - 1) ? 0 : (closest + 1);
+		double leftDist = Location.distance(l, locs[left]);
+		double rightDist = Location.distance(l, locs[right]);
+		// Iterate over at most m closest peers
+		while (left != right) {
+			if (leftDist <= rightDist) {
+				final double loc = locs[left];
+				if (!exclude.contains(loc)) {
+					return loc;
+				}
+				left = (left == 0) ? (locs.length - 1) : (left - 1);
+				leftDist = Location.distance(l, locs[left]);
+			} else {
+				final double loc = locs[right];
+				if (!exclude.contains(loc)) {
+					return loc;
+				}
+				right = (right == locs.length - 1) ? 0 : (right + 1);
+				rightDist = Location.distance(l, locs[right]);
+			}
+		}
+		final double loc = locs[left];
```

Comment:
this is worst case performance, so it might be useful to add some logging here.

The average runtime of this algorithm should not be *O(log (n) + m)* but rather something slightly above *O(log(n) + m/(n-m))*, since the exclude set is likely to be distributed randomly. It would be O(log(n) + m/(n-m)) for a uniform distribution of exclude in the location list.